engine/execution/state/state.go

```diff
@@ -36,8 +39,15 @@ type ReadOnlyExecutionState interface {
 // ScriptExecutionState is a subset of the `state.ExecutionState` interface purposed to only access the state
 // used for script execution and not mutate the execution state of the blockchain.
 type ScriptExecutionState interface {
-	// NewStorageSnapshot creates a new ready-only view at the given state commitment.
-	NewStorageSnapshot(flow.StateCommitment) snapshot.StorageSnapshot
+	// NewStorageSnapshot creates a new ready-only view at the given block.
+	NewStorageSnapshot(commit flow.StateCommitment, blockID flow.Identifier, height uint64) snapshot.StorageSnapshot
+
+	// CreateStorageSnapshot creates a new ready-only view at the given block.
+	// It returns:
+	// - (nil, nil, storage.ErrNotFound) if block is unknown
+	// - (nil, nil, state.ErrNotExecuted) if block is not executed
+	// - (nil, nil, state.ErrExecutionStatePruned) if the execution state has been pruned
+	CreateStorageSnapshot(blockID flow.Identifier) (snapshot.StorageSnapshot, *flow.Header, error)
```

Comment:
Do we still need the first method? 
I wonder (not sure the impact of the change on the current usage of the method) if we could have one method that you could call with options, something like:
```
CreateStorageSnapshot(WithBlockID(id)) 
CreateStorageSnapshot(WithCommitment(commitment), WithBlockID(id), WithHeight(height)) 
```
You could then internally decide how to handle that. I don't have much context on the usage tho, so feel free to disregard

Reply:
There are a few different scenarios that still need the `NewStorageSnapshot` function. Let me explain why we need these two methods:

1. There are two storage engines, ledger-based storage (current) and storehouse-based storage (upgrading to). Ledger-based storage takes commitment to find register, whereas storehouse takes BlockID and Height. We would like to use a flag to control which storage engine to use in the execution state, and we don't want the caller to know this implementation detail. That's why we pass all three data to the NewStorageSnapshot method, so that the execution state can decide internally which data to use. If we make them as options like what you suggested with `WithCommitment/WithBlockID`, then it means the caller still has to know the implementation detail to decide which option to use.

2. Usually, we will lookup block first before querying the register, that's why I made this function `CreateStorageSnapshot` that takes a blockID, and lookup the commitment or block height and then create the storage snapshot. 

3. But then, why do we still need the NewStorageSnapshot, if we always need lookup commitment or block height ? Because there are one scenario where we don't need to lookup commitment or block height. That's when we are executing a block, the commitment and the block height is already available, [see this change in ingestion engine](https://github.com/onflow/flow-go/pull/5031/files#diff-e5c0dea70cf688717060471583da13cc5269a6f661e955613ed7ef698354cc1aR432). Using CreateStorageSnap there would make a unnecessary call.